Add private named parameters section (#7250)
dart-lang/site-www: src/content/language/constructors.md

## Instance variable initialization

Dart provides several ways to initialize instance variables.
You can assign values in the declaration,
use initializing formal parameters,
or use an initializer list.
For private fields,

> Dart also supports a dedicated syntax using private named parameters.

Reviewer: I'm not sure if this last sentence is worth calling out here. It's not really a separate feature so much as if you have an initializing formal that happens to be a named parameter and have a private name, then it does this little extra thing where the parameter gets a public name.

> Dart also supports a dedicated syntax using private named parameters.

Reviewer 2: I like the expansion of this introduction!

However, I think this additional mention might not be needed as it's just a part of (the already mentioned) initializing formal parameters support.

}
```

> ### Private named parameters

Reviewer 2: Should we name this section and references to this capability as "private-named parameters" to clarify the parameter itself isn't private, but rather the name is?

Author: @munificent thoughts?

Reviewer: It's not really just about parameters whose name happens to be private. Dart has already allowed those for positional parameters:

```dart
class C {
  int _x;
  C(this._x);
}
```

It's really about named parameters whose name is private. I thought about calling the feature "private-named named parameters" but that's obviously no fun for anyone to read or say. :) For what it's worth, I think of the feature as "private (named parameters)" more than "(private named) parameters".

But if you think the hyphen will help clarify things for readers, I'm OK with it.

Author: I think we should keep the title as is to remain consistent with the feature spec (if that matters). Do you have strong opinions about the hyphen @parlough?

Reviewer 2: Nope, I'm happy with keeping it as is and being consistent! Thanks for considering this though :)

:::version-note
Using private named parameters as initializing formals
requires a [language version][] of at least 3.12.
:::

In Dart, fields starting with an underscore are private to their library.
To initialize a private field using a named parameter,
you previously had to write manual assignment boilerplate

> in the initializer list:

Reviewer 2: I think this framing is fine, but eventually "previously" will get a bit out dated, with future readers not writing code before the new syntax existed. Optionally consider adjusting this to a "you can" structure or leaving a comment to revisit this in the future.

<?code-excerpt "point_private_old.dart (initialize-private-named-before)" plaster="none"?>
```dart
class Point {
  final double _x;
  Point({required double x}) : _x = x;
}
```

You can initialize private fields directly
in the constructor parameter list.
When you prefix the named parameter with `this._`,
the compiler automatically strips the underscore for the caller,
allowing them to use a clean, public name:

<?code-excerpt "point_private_new.dart (initialize-private-named-after)" plaster="none"?>
```dart
class Point {
  final double _x;
  Point({required this._x});
}
```

In both cases,
the caller uses the public name `x` at the call site:

<?code-excerpt "point_private_new.dart (initialize-private-named-usage)" plaster="none"?>
```dart
var p = Point(x: 1.0);
```

> Like regular named parameters, you can make private named parameters optional or required.

Reviewer 2: Optionally consider cross linking named parameters to https://dart.dev/language/functions#named-parameters.

> Like regular named parameters, you can make private named parameters optional or required.

Reviewer 2: ```suggestion
Like regular named parameters, you can
make private named parameters optional or required.
```

You can also provide explicit default values.

In the following example, the `_x` parameter is optional and defaults to `null`.

> The `_y` parameter is also optional but uses an explicit default value of `0.0`:

Reviewer 2: ```suggestion
The `_y` parameter is also optional but has an explicit default value of `0.0`:
```

<?code-excerpt "point_alt.dart (initialize-private-named)" plaster="none"?>
```dart
class PointPrivate {
  final double? _x; // Nullable field
  final double _y;  // Non-nullable field

  PointPrivate({this._x, this._y = 0.0});

  @override
  String toString() => 'PointPrivate($_x, $_y)';
}

void testPrivate() {
  var p = PointPrivate(x: 1.0, y: 2.0);
  print(p);
}

> ```

Reviewer: These examples are all excellent!

#### Constraints

* **No conflicts:** Neither the private name nor the generated public name
  can match any other parameter name in the same constructor.
* **Initializing formals only:** Named parameters in Dart generally

>   cannot be private. This feature is an exception that only applies

Reviewer 2: Consider changing "feature" to "capability". It seems less ambiguous and more consistent for the future, where this "new feature" just becomes a part of the language.

>   to named parameters that are initializing formals (`this._field`).

Reviewer: Just a note for the future: When primary constructors arrive in Dart 3.13, you will also be able to use private named parameters for declaring parameters like:

```dart
class Point({final int _x = 0});
```

So it's correct now that it's only for initializing formals, but for Dart 3.13, it will be initializing formals and declaring parameters. :)

>   You cannot use private names for regular named parameters.

Reviewer 2: Generally we use "can't" and similar for [negation contractions](https://developers.google.com/style/contractions#negation-contractions).

```suggestion
  can't be private. This feature is an exception that only applies
  to named parameters that are initializing formals (`this._field`).
  You can't use private names for regular named parameters.
```

>   You cannot use private names for regular named parameters.

Reviewer 2: Maybe "identifiers" is more accurate than "names"? I'm not sure, but it seems like a name can't be private but an identifier(/something with an identifier) can be.

> * **Valid public name:** The private name must map to a valid public identifier.

Reviewer 2: Good call out!

  For example, `this._` or `this._2x` are invalid
  because they don't have valid public counterparts.

#### Usage in initializer lists

Within the constructor's initializer list,

> you refer to the parameter using its private name:

Reviewer 2: ```suggestion
reference the parameter using its private name:
```

<?code-excerpt "point_alt.dart (initialize-private-named-assert)" plaster="none"?>
```dart
class PointPrivateAssert {
  final double _x;

  PointPrivateAssert({required this._x}) : assert(_x >= 0);
}
```

#### Interaction with super parameters

When extending a class that uses private named parameters,

> subclasses use the public name for [super parameters](/resources/glossary#super-parameter).

Reviewer 2: Consider using a link reference for this and declaring the definition at the end of the section.

```
subclasses use the public name for [super parameters][].

[super parameters]: /resources/glossary#super-parameter
```

In the following example, the `Tool` class defines the private field `_price`.

> Even though the field is private, the compiler exposes the parameter name publicly.

Reviewer 2: ```suggestion
Even though the field is private,
the compiler exposes the parameter name publicly.
```

> Even though the field is private, the compiler exposes the parameter name publicly.

Reviewer 2: I'm not sure a better way to word this, but to me, "the compiler" being the actor here seems a bit complicated or not fully accurate. It's a language feature. It's also not clear to me until the next sentence what "exposing a parameter name publicly" exactly means.

In absence of a better actor, would it be fine to just say the parameter is public? Something like the following:

```
Even though the field is private,
its corresponding named parameter is public (`price` not `_price`).
```
